### PR TITLE
Assign a UUID to Each AST Node

### PR DESCRIPTION
**Pull request by Josef-Vonasek**
### Pull Request Description

Assigns an UUID to each AST node but Module.
Changes Infix.opr to `T`, so that it is correctly assigned an UUID from idmap.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/style-guides/scala.md), [Java](https://github.com/luna/enso/blob/master/doc/style-guides/java.md), [Rust](https://github.com/luna/enso/blob/master/doc/style-guides/rust.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/style-guides/haskell.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/syntax/definition/src/main/scala/org/enso/data/Index.scala [hunk 1]
```diff
@@ -17,4 +17,6 @@ case class Index(value: Int) extends AnyVal with Ordered[Index] {
 
 object Index {
   val Start = Index(0)
+
+  def apply(): Index = Index(0)
```
Comment (kustosz) [inline lib/syntax/definition/src/main/scala/org/enso/data/Index.scala:21]: This seems very strange. Why would `Index()` mean anything. You already have `Index.Start` precisely for that use case.

File: lib/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala [hunk 2]
```diff
@@ -722,8 +722,10 @@ object Shape extends ShapeImplicit {
     implicit def repr[T: Repr]: Repr[Infix[T]] =
       t => R + t.larg + t.loff + t.opr + t.roff + t.rarg
     implicit def ozip[T: HasSpan]: OffsetZip[Infix, T] = t => {
-      val rargIndex = Index(t.larg.span + t.loff + t.opr.span + t.roff)
-      t.copy(larg = (Index.Start, t.larg), rarg = (rargIndex, t.rarg))
+      val larg = Index()                                           -> t.larg
```
Comment (kustosz) [inline lib/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala:725]: `
```suggestion
      val larg = Index.Start                                           -> t.larg
```

File: lib/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala [hunk 3]
```diff
@@ -389,8 +387,10 @@ class Parser {
       ozip: OffsetZip[T, AST]
     ) {
       def process(context: Context): AST.ASTOf[T] = {
-        def assignmentInBlock(opr: AST.Opr): Boolean =
-          context == Block && opr.name == "="
+        def assignmentInBlock(opr: AST._AST): Boolean = opr match {
+          case AST.Ident.Opr.any(opr) if context == Block && opr.name == "=" => true
+          case _                                                             => false
+        }
```
Comment (farmaazon) [inline lib/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala:393]: Do we still need this function? I believe that it was used to assign ids to nodes, but now we assign ids everywhere.

File: lib/syntax/definition/src/main/scala/org/enso/data/Index.scala (hunk 1 above)
Comment (iamrecursion) [reply lib/syntax/definition/src/main/scala/org/enso/data/Index.scala:21]: Agreed.

File: lib/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala [hunk 4]
```diff
@@ -389,8 +387,10 @@ class Parser {
       ozip: OffsetZip[T, AST]
     ) {
       def process(context: Context): AST.ASTOf[T] = {
-        def assignmentInBlock(opr: AST.Opr): Boolean =
-          context == Block && opr.name == "="
+        def assignmentInBlock(opr: AST._AST): Boolean = opr match {
```
Comment (iamrecursion) [inline lib/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala:390]: I'm not comfortable giving up the guarantee that an operator has a valid operator name.

File: lib/syntax/definition/src/main/scala/org/enso/data/Index.scala (hunk 1 above)
Comment (Josef-Vonasek) [reply lib/syntax/definition/src/main/scala/org/enso/data/Index.scala:21]: Well, why shouldn't it mean something? I read `Type()` as `Type.DEFAULT`.

If you find the original version more understandable, then I don't mind changing i back..

Comment (kustosz) [reply lib/syntax/definition/src/main/scala/org/enso/data/Index.scala:21]: Correct, `Type()` means `Type.Default`. My point is: `Index` is not a type that has a meaningful default.

Comment (kustosz) [reply lib/syntax/definition/src/main/scala/org/enso/data/Index.scala:21]: Which is also why you have `Index.Start` not `Index.Default`.

File: lib/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala (hunk 3 above)
Comment (Josef-Vonasek) [reply lib/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/Parser.scala:393]: Good catch!

**Pull request merged**